odoo/addons/base/tests/test_ir_mail_server_smtpd.py

```diff
@@ -0,0 +1,443 @@
+
+from odoo.exceptions import UserError
+from odoo.tools import file_path, mute_logger
+from .common import TransactionCaseWithUserDemo
+
+try:
+    import aiosmtpd
+    import aiosmtpd.controller
+    import aiosmtpd.smtp
+    import aiosmtpd.handlers
+except ImportError:
+    aiosmtpd = None
+
+
+PASSWORD = 'secretpassword'
+_openssl = shutil.which('openssl')
+_logger = logging.getLogger(__name__)
+
+
+def _find_free_local_address():
+    """ Get a triple (family, address, port) on which it possible to bind
+    a local tcp service. """
+    addr = aiosmtpd.controller.get_localhost()  # it returns 127.0.0.1 or ::1
+    family = socket.AF_INET if addr == '127.0.0.1' else socket.AF_INET6
+    with socket.socket(family, socket.SOCK_STREAM) as sock:
+        sock.bind((addr, 0))
+        port = sock.getsockname()[1]
+    return family, addr, port
+
+
+def _smtp_authenticate(server, session, enveloppe, mechanism, data):
+    """ Callback method used by aiosmtpd to validate a login/password pair. """
+    result = aiosmtpd.smtp.AuthResult(success=data.password == PASSWORD.encode())
+    _logger.debug("AUTH %s", "successfull" if result.success else "failed")
+    return result
+
+
+class Certificate:
+    def __init__(self, key, cert):
+        self.key = key
+        self.cert = cert
+
+    def __repr__(self):
+        return f"Certificate({self.key=}, {self.cert=})"
+
+
+def _generate_certificates(certdir):
+    """Generate and sign a few RSA key pairs inside certdir.
+
+    It invokes the openssl command line to create a new Certification
+    Authority and then use that CA to generate extra keys and
+    certificates for a user, a server, and any purpose.
+
+    It also creates a self-signed certificate, to test that invalid
+    certificates are correctly rejected.
+
+    It returns 4 RSA: CA, web server, mail server, self-signed. Each RSA
+    is a (named)tuple of 2 elements: key and cert.
+    """
+    # Create a fake certificate authority and use it to generate a few
+    # signed certificates.
+    shutil.copy(file_path('base/tests/ssl/openssl.conf'), certdir)
+    shutil.copy(file_path('base/tests/ssl/gencert.sh'), certdir)
+    subprocess.run(
+        [shutil.which('sh'), certdir / 'gencert.sh'],
+        check=True,
+        cwd=certdir,
+        stdout=sys.stdout if _logger.isEnabledFor(logging.DEBUG) else subprocess.DEVNULL,
+        stderr=subprocess.STDOUT,
+    )
+
+    # Collect files and return them
+    return (
+        Certificate(key=None, cert=certdir / 'ca.cert.pem'),
+        Certificate(key=certdir / 'client.key.pem', cert=certdir / 'client.cert.pem'),
+        Certificate(key=certdir / 'server.key.pem', cert=certdir / 'server.cert.pem'),
+        Certificate(key=certdir / 'self_signed.key.pem', cert=certdir / 'self_signed.cert.pem'),
+    )
+
+
+# skip when optional dependencies are not found
+@unittest.skipUnless(aiosmtpd, "aiosmtpd couldn't be imported")
```

Comment:
Is it available on runbot, aka will that be tested on runbot ?

Reply:
aiosmtpd is installed in runbot Docker images used for tests. So it will work as expected.
What we don't want is to have dependency in the requirements. So it's better to skip in order to avoid crashes in local environments where aiosmtpd is not installed.